### PR TITLE
Adapt unsolvable feature in CFI.

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**
Adapt PR for upstream change in https://github.com/opprop/checker-framework-inference/pull/128

### CODE REVIEW

File: src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java [hunk 1]
```diff
@@ -111,12 +113,16 @@ private void addPreferenceToCurBottom(ConstantSlot curBtm, Set<Constraint> consS
...
         Map<Integer, EnumSet<OntologyValue>> ontologyResults = new HashMap<> ();
 
-        for (Map<Integer, AnnotationMirror> inferenceSolutionMap : inferenceSolutionMaps) {
-            for (Map.Entry<Integer, AnnotationMirror> entry : inferenceSolutionMap.entrySet()) {
+        for (Pair<Map<Integer, AnnotationMirror>, Collection<Constraint>> inferenceResult : inferenceResults) {
+            if (inferenceResult.fst == null) {
+                return new DefaultInferenceResult(inferenceResult.snd);
+            }
+
+            for (Map.Entry<Integer, AnnotationMirror> entry : inferenceResult.fst.entrySet()) {
```
Comment (topnessman) [inline src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java]: Can you put it into else branch to make it more clear?

File: src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java [hunk 2]
```diff
@@ -142,6 +148,6 @@ protected InferenceSolution mergeSolution(List<Map<Integer, AnnotationMirror>> i
             result.put(entry.getKey(), resultAnno);
         }
 
-        return new DefaultInferenceSolution(result);
+        return new DefaultInferenceResult(result);
```
Comment (topnessman) [inline src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java]: `result` -> `solutions`

File: src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java (hunk 1 above)
Comment (topnessman) [reply src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java]: Importing `Map.Entry` and using `Entry` directly would be better.

File: src/ontology/util/OntologyStatisticUtil.java [hunk 3]
```diff
@@ -150,7 +150,7 @@ private static String getHashCode(OntologyValue... values) {
      * @param constraints
      * @param qualifierHierarchy
      */
-    public static void verifySolution(InferenceSolution mergedSolution,
+    public static void verifySolution(InferenceResult mergedSolution,
```
Comment (topnessman) [inline src/ontology/util/OntologyStatisticUtil.java]: `mergedSolution` -> `mergedResult`

Comment (topnessman) [reply src/ontology/util/OntologyStatisticUtil.java]: Is `verifySolution()` only called if there is solution? If yes, adding `assert mergedResult.hasSolution()` would be good. If `verifySolution()` is accidentally called if no solution, calling `mergedResult.getSolutionForVariable()` throws NPE.

Comment (CharlesZ-Chen) [reply src/ontology/util/OntologyStatisticUtil.java]: yes, `verifySolution` should be called if there is a solution. Let me add the assertion.

File: src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java (hunk 1 above)
Comment (CharlesZ-Chen) [reply src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java]: > Can you put it into else branch to make it more clear?

As there is a return statement in the `if` branch, it is obvious that line 125 only can be reached in the else branch. Adding an explicit else make the for loop further nested, and seems a bit not easy to read....

Comment (topnessman) [reply src/ontology/solvers/backend/OntologyGraphSolvingStrategy.java]: Hmm, then it's ok to keep it.

**Comment by CharlesZ-Chen** [comment]
@topnessman Comments addressed.

**Pull request merged**